src/client/Microsoft.Identity.Client/Http/Headers/WwwAuthenticateParameters.cs

```diff
@@ -0,0 +1,464 @@
+        /// <param name="resourceUri">URI of the resource.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel operation.</param>
+        /// <returns>WWW-Authenticate Parameters extracted from response to the unauthenticated call.</returns>
+        public static async Task<IReadOnlyList<WwwAuthenticateParameters>> CreateFromAuthenticationResponseAsync(HttpClient httpClient, string resourceUri, CancellationToken cancellationToken = default)
+        {
+            if (httpClient is null)
+            {
+                throw new ArgumentNullException(nameof(httpClient));
+            }
+            if (string.IsNullOrWhiteSpace(resourceUri))
+            {
+                throw new ArgumentNullException(nameof(resourceUri));
+            }
+
+            // call this endpoint and see what the header says and return that
+            HttpResponseMessage httpResponseMessage = await httpClient.GetAsync(resourceUri, cancellationToken).ConfigureAwait(false);
+            var wwwAuthParams = CreateFromAuthenticateHeaders(httpResponseMessage.Headers);
+            return wwwAuthParams;
+        }
+
+        /// <summary>
+        /// Create WWW-Authenticate parameters from the HttpResponseHeaders for each auth scheme.
+        /// </summary>
+        /// <param name="httpResponseHeaders">HttpResponseHeaders.</param>
+        /// <returns>The parameters requested by the web API.</returns>
+        /// <remarks>Currently it only supports the Bearer scheme</remarks>
+        public static IReadOnlyList<WwwAuthenticateParameters> CreateFromAuthenticateHeaders(
+            HttpResponseHeaders httpResponseHeaders)
+        {
+            List<WwwAuthenticateParameters> parameterList = new List<WwwAuthenticateParameters>();
+
+            foreach (AuthenticationHeaderValue wwwAuthenticateHeaderValue in httpResponseHeaders.WwwAuthenticate)
+            {
+                var parameters = CreateFromWwwAuthenticateHeaderValue(wwwAuthenticateHeaderValue.Parameter);
+                parameters.AuthScheme = wwwAuthenticateHeaderValue.Scheme;
+                parameterList.Add(parameters);
+            }
+
+            return parameterList;
+        }
+        #endregion Multi Scheme Api
+
+        /// <summary>
+        /// Creates parameters from the WWW-Authenticate string.
+        /// </summary>
+        /// <param name="wwwAuthenticateValue">String contained in a WWW-Authenticate header.</param>
+        /// <returns>The parameters requested by the web API.</returns>
+        public static WwwAuthenticateParameters CreateFromWwwAuthenticateHeaderValue(string wwwAuthenticateValue)
+        {
+            if (string.IsNullOrWhiteSpace(wwwAuthenticateValue))
+            {
+                throw new ArgumentNullException(nameof(wwwAuthenticateValue));
+            }
+            IDictionary<string, string> parameters;
+
+            var authValuesSplit = wwwAuthenticateValue.Split(new char[] { ' ' }, 2);
```

Comment:
Per [spec](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate#syntax) the value can have multiple challenges (ex. bearer, pop) separated by comma. So we should split by `,` and this method should return a collection of WwwAuthenticateParameters really.

Reply:
Actually, this method is also supposed to be obsolete. Instead of allowing developers to pass in any string they want, we are now going to only allow them to pass in the `Headers` collection from the .NET http response. .NET will automatically seperate headers by `'` and create a list of challenges that we can parse individually.